### ISSUE TITLE
Issue with dotnet restore on NET8 and multi nuget source configured.

### ISSUE THREAD

**Issue by rainersigwald** [NuGet/Home#13129]
---
**Issue moved from dotnet/core#9012**
- Please respond to @DmitryLukyanov.

---

_From @DmitryLukyanov on Monday, December 18, 2023 6:21:56 PM_

## Description

Consider you have multi nuget source configured on the machine in nuget.config:

     <packageSources>
          <add key="nuget.org" value=https://api.nuget.org/v3/index.json protocolVersion="3" />
          <add key="externalSourceThatRequireAuth" value="https://.../nuget/v3/index.json"/>
          <add key="Microsoft Visual Studio Offline Packages" value="C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\" />
    </packageSources>

Then try to call `dotnet restore` as described below:

    c:\>mkdir issue
    c:\>cd issue
    c:\issue>mkdir app
    c:\issue>cd app
    c:\issue\app>dotnet --version
    8.0.100
    c:\issue\app>dotnet new console --no-restore
    The template "Console App" was created successfully.
    c:\issue\app>dotnet restore app.csproj
      Determining projects to restore...
    C:\Program Files\dotnet\sdk\8.0.100\NuGet.targets(156,5): error : Unable to load the service index for source https://.../nuget/v3/index.json. [c:\issue\app\app.csproj]
    C:\Program Files\dotnet\sdk\8.0.100\NuGet.targets(156,5): error :   Response status code does not indicate success: 401 (Unauthorized). [c:\issue\app\app.csproj]

Notice that I don't use `externalSourceThatRequireAuth` in this app. Then there are 3 workarounds to fix it:
1. Call something like `nuget search -take 10 -Source externalSourceThatRequireAuth` that will solve issue approximately for 30 mins.
2. Use `msbuild` instead `dotnet`: 

        c:\issue\app>msbuild app.csproj -t:Restore

4. switch back to older target frameworks, I checked with .NET5:

        c:\issue\app>cd ../    
        c:\issue>dotnet --list-sdks
        5.0.408 [C:\Program Files\dotnet\sdk]
        8.0.100 [C:\Program Files\dotnet\sdk]     
        c:\issue>dotnet new globaljson
        The template "global.json file" was created successfully.    
        c:\issue>more global.json
        {
          "sdk": {
              "version": "8.0.100"
           }
        }     
// change the sdk version on 5.0.408      

    c:\issue>more global.json
    {
      "sdk": {
        "version": "5.0.408"
      }
    }     
    c:\issue>cd app     

// change the version of the project from `net8.0` to `net5.0`

    c:\issue\app>more app.csproj
    ...
        <TargetFramework>net5.0</TargetFramework>
    ...     
    c:\issue\app>dotnet restore app.csproj
      Determining projects to restore...
      Restored c:\issue\app\app.csproj (in 63 ms).

### Configuration

* Which version of .NET is the code running on: net8.0
* What OS and version: windows 11
* What is the architecture: x64

### Regression?

yes, it works well on .net5.0 and I think it worked on .net7.0 (didn't check now though)

**Comment by rainersigwald** [comment]
---
**Issue moved from dotnet/core#9012**
- Please respond to @wfurt.

---

_From @wfurt on Thursday, January 4, 2024 7:06:06 PM_

any idea @rainersigwald ? If now who should look at this? 
It seems like `restore` should simply ignore sources that cannot be access.

**Comment by rainersigwald** [comment]
---
**Issue moved from dotnet/core#9012**
- Please respond to @rainersigwald.

---

_From @rainersigwald on Thursday, January 4, 2024 8:37:14 PM_

Let's move this to NuGet, just a moment.

**Comment by jeffkl** [comment]
@DmitryLukyanov can you please try specifying `--interactive` when doing `dotnet new`?  That should give you an authentication prompt.  If your feed is Azure DevOps, you might need the latest authentication provider (https://github.com/microsoft/artifacts-credprovider) or to configure it to prompt you.

**Comment by DmitryLukyanov** [comment]
> @DmitryLukyanov can you please try specifying `--interactive` when doing `dotnet new`? That should give you an authentication prompt. If your feed is Azure DevOps, you might need the latest authentication provider (https://github.com/microsoft/artifacts-credprovider) or to configure it to prompt you.

`--interactive` - didn't give any difference. 

> If your feed is Azure DevOps

yes, I use Azure DevOps

>you might need the latest authentication provider (https://github.com/microsoft/artifacts-credprovider) or to configure it to prompt you.

it might be the solution (I haven't tried), but as I can see, there are enough evidences that it's more bug than configuration issue. Why msbuild works without issues? Why no issues with older TFs? Why if I query `nuget search` the issue is temporary resolved without any prompts? Regarding the latest, I use this as workaround, if my script fails, I simply fetch nugets I have and then approximately next 30 minutes it's working well, then.. try again :(

**Comment by jeffkl** [comment]
NuGet itself does not provide functionality to authenticate with feeds other than a password stored in your NuGet.config file.  To authenticate with a particular feed, a plug-in is required that NuGet calls to acquire an auth token for that specific feed.  By default, a "build" which calls the restore target is considered non-interactive and will not prompt for authentication.  Its hard to say with the limited information given if the auth provider is being called at all or just not prompting for authentication.  

Other commands like `nuget search` are interactive by default and is probably calling the Azure DevOps auth provider for you which appears seamless.  I use Azure DevOps extensively and have no issues.  But we'll need to figure out why your restore invocation is not getting an auth token from the credential provider.

**Comment by DmitryLukyanov** [comment]
still, it doesn't explain the difference between behavior in `.net8` and previous `TF`s :) 

> Its hard to say with the limited information given 

let me know what information you need

**Comment by jeffkl** [comment]
> still, it doesn't explain the difference between behavior in .net8 and previous TFs :)

There might have been changes in the Azure Artifacts Credential Provider in those releases.  

> let me know what information you need

1. Have you tried any of the troubleshooting steps at https://github.com/microsoft/artifacts-credprovider?tab=readme-ov-file#use ?
2. Have you looked at the verbose output of a restore when specifying `--interactive` to see if there's additional information?  By default, `dotnet restore` doesn't show very much.  You can specify `--verbosity normal` or `--verbosity detailed`

**Comment by DmitryLukyanov** [comment]
> > still, it doesn't explain the difference between behavior in .net8 and previous TFs :)
> 
> There might have been changes in the Azure Artifacts Credential Provider in those releases.
> 
> > let me know what information you need
> 
>     1. Have you tried any of the troubleshooting steps at https://github.com/microsoft/artifacts-credprovider?tab=readme-ov-file#use ?
> 
>     2. Have you looked at the verbose output of a restore when specifying `--interactive` to see if there's additional information?  By default, `dotnet restore` doesn't show very much.  You can specify `--verbosity normal` or `--verbosity detailed`

>Have you looked at the verbose output 

yes, there was nothing interesting there as far as I can tell, I will try to do it again in some time and will go back with details

>Have you tried any of the troubleshooting steps at

I didn't. I don't really understand why I have to do it since the only change I did is replacing used .net version on latest in my application. From what I can tell, even though it may help, it should be a BC issue somewhere in `dotnet restore` logic (msbuild works fine)

**UPDATE:**
>Have you tried any of the troubleshooting steps at

I missed that there are some troubleshooting steps there. At first glance, it looks unrelated, but I will look deeply in some time

**Comment by anluoma** [comment]
We are also hitting this after update from net 6 -> net 8. We do use AzureDevOps (server) as well.

**Comment by brunoquinelato** [comment]
I encountered the same issue you reported on Azure Pipelines while utilizing .NET 8. Upon reviewing responses from [jeffkl](https://github.com/jeffkl), I identified a resolution by incorporating the NuGetAuthenticate@1 task prior to executing the 'dotnet restore' command.

![image](https://github.com/NuGet/Home/assets/69924232/ce0f8595-5638-4e46-8b91-ec926fe29cc9)

**Comment by anluoma** [comment]
thanks @brunoquinelato - we added nugetAuth task ( we had it on other templates but not these it seems). but no luck with NET8. 
Also NET 6 started to fail. 

Maybe (not sure anymore) after NET 6 build / restarting build agents NET 6 seems to work again. 
Something is caching something. 

From the logs I can see that NuGetAuthenticate task is saying that "Setting up the credential provider to use the identity..." but it lists only the http version of the URL not **https** which is is the what we have in nuget.config so I expect that it would show both urls when saying about that credential provider..

**Comment by anluoma** [comment]
Added env VSS_NUGET_EXTERNAL_FEED_ENDPOINTS for DotNetCoreCLI@2 task. Which seems to resolve issue, allthough this is internal feed.

**Comment by cermakp** [comment]
Has it been somehow resolved? It seems that the problem is still there and there is a connection with .NET SDK 8.x. In previous version there was no issue with authentication when using nuget restore. Seems that the problem can be solved by installation of the latest version of [Azure Artifact Credential Provider](https://github.com/microsoft/artifacts-credprovider). However I would expect that this extra installation will not be needed in case that we will start using .NET SDK 8.x

**Comment by marlon-tucker** [comment]
I am running into this issue on our build infrastructure which uses TeamCity, the build agents are just virtual machines.

We have added azure devops PAT tokens into the build users Nuget.config for the custom feed with the PAT in the `cleartextpassword` property.

This setup has historically been fine, however in the last two weeks this is now giving us issues where dotnet is not able to authenticate with the devops feed (returns 401).

What is strange about the situation is if I use `Nuget.exe` to perform the restore, it works, but using `dotnet restore` results in the 401 failure.

I've attempted to install the credentials plugin and setup the `VSS_NUGET_EXTERNAL_FEED_ENDPOINTS` environment variable but even that does not seem to be allowing `dotnet` to successfully restore.

I'm at a bit of a loss how to debug this further really.
